lib/plugins/cors.js

```diff
@@ -73,66 +120,234 @@ function matchOrigin(req, origins) {
  * Pre-flight requests are handled by the router internally
  *
  * @public
- * @function cors
- * @param    {Object}   options an options object
- * @returns  {Function}
+ * @function createCorsContext
+ * @param    {Object} options an options object
+ * @param    {Array} [options.origins] an array of whitelisted origins
+ * @param    {Boolean} [options.credentials] if true, uses creds
+ * @param    {Array} [options.allowHeaders] user defined headers to allow
+ * @param    {Array} [options.exposeHeaders] user defined headers to expose
+ * @param    {Array} [options.preflightMaxAge] ms to cache preflight requests
+ * @param    {Object | Function} [options.preflightStrategy]
+ * customize preflight request handling
+ * @returns  {Object} returns an object with a handler and preflight handler
  */
-function cors(options) {
-    assert.optionalObject(options, 'options');
-    var opts = options || {};
-    assert.optionalArrayOfString(opts.origins, 'options.origins');
-    assert.optionalBool(opts.credentials, 'options.credentials');
-    assert.optionalArrayOfString(opts.headers, 'options.headers');
+function createCorsContext(options) {
 
-    cors.credentials = opts.credentials;
-    cors.origins = opts.origins || ['*'];
+    assert.object(options, 'options');
+    assert.optionalBool(options.credentials, 'options.credentials');
+    assert.optionalArrayOfString(options.exposeHeaders,
+                                 'options.exposeHeaders');
+    assert.optionalArrayOfString(options.allowHeaders, 'options.allowHeaders');
+    assert.optionalNumber(options.preflightMaxAge, 'options.preflightMaxAge');
 
-    var headers = (opts.headers || []).slice(0);
-    var origins = opts.origins || ['*'];
+    var opts = options || {};
+    var credentials = opts.credentials || false;
+    // convert all origin values into regexps. however, leave '*' as is for
```

Comment:
It will be inevitably special cased because everything gets converted to a regex with http/https in front of it, and we have to figure out what http://* means. No?